mathics/packages/DiscreteMath/Kernel/init.m

```diff
@@ -11,6 +11,7 @@
  *)
 Get[ "DiscreteMath`CombinatoricaV0.6" ]
 Get[ "DiscreteMath`CombinatoricaV0.9" ]
```

Comment:
Shouldn't only one version be loaded?

Reply:
I think you are correct. Also we are loading in RSolve as well which is different. 

I'll mark this as a draft then. 

I don't understand how Mathics packaging works. What is intended is to load the package based on the name that is given in "Needs". For example if I write ``Needs["DiscreteMath`CombinatoricaV2.0`"]`` then only V2.0 should get loaded. But if I type ``Needs["DiscreteMath`CombinatoricaV0.6`"]`` then only V0.6 should get loaded. 

Also, when I try test_combinatorica (for version 0.9) using the 2.0 version there is some code that fails. So right now 0.9 may work a little better for now. 

Here are some other things that might be a bug. 

Here are a couple of messages that I get when loading this unmodified: 

```
In[1]:= Needs["DiscreteMath`CombinatoricaV2.0`"]
Set::write: Tag All in All::usage is Protected.
Syntax::com: Warning: comma encountered with no adjacent expression. The expression will be treated as Null (line 1266 of "/src/external-vcs/github/Mathics3/mathics-core/mathics/packages/DiscreteMath/CombinatoricaV2.0.m").
```

With respect to the `Set::write` message the system is basically saying that `All:usage` has already been defined which is why it is protected. 

However the code has:

```
BeginPackage["DiscreteMath`Combinatorica2`"]
...
All::usage = "All is an option to inform certain functions to return all solutions, instead of just the first one."
```

So I would have expected `All` to be in the ``DiscreteMath`Combinatorica2`` namespace rather than ``System`All`.

As for: `Syntax::com: Warning: comma encountered with no adjacent expression.`, 

we have: 

```
		For [j=2, (j<n)  && (y[[j]]>=y[[j-1]]), j++, ];
```

and it doesn't like the trailing comma followed by nothing. 

If you change that to: 

```
		For [j=2, (j<n)  && (y[[j]]>=y[[j-1]]), j++, Null];
```

then the warning goes away. 

@mmatera should there be a warning?  And is there a bug in the namespacing in Mathics regarding the two "All"s

Reply:
Regarding `All`, we should check if it is not a bug of the package: it is possible that in an earlier version of WMA `All` were not a system symbol, or maybe it was not protected. On the other hand, it could happen that the `BeginPackage` mechanism in Mathics has a bug.
 
Regarding the `For` loop, in WMA, I got the warning: this is because the "body" of the `For` loop is missing.

Reply:
So for the `For` loop, I will add back in `Null` since that is preferred. 

For `All`,  the question in my mind is less about whether WMA `All` was defined at the time the package was written but rather whether `BeginPackage` should allow a symbol that had been defined in `System` to get defined in the context of the package without getting that protected message.

If we wanted to handle the situation that `All` wasn't defined in Mathematica version 5 but is in Version 6, then we would write a Version 5 compatibility package that removes `All`.

Reply:
@rocky: regarding the warning on setting ``All::usage``, it seems that in WMA, it is possible to assign messages to protected symbols. I am going to add a PR changing that here.

Reply:
Ok - thanks. However the question I have is about the namespace aspect. Is setting `All::usage` setting ``System`All::usage`` or ``Combinatorica2`All::usage``?    If this is the case, the bug feels more in the category of Mathics not separating namespaces correctly.

Reply:
@rocky, the answer to this, as well the fix for the assignment module are in the PR #283.

Inside a Begin/End block or in a BeginPackage/EndPackage block, symbols in the context `"System`"` are accessible, so
```
BeginPackage["acontext`"]
None::usage= "someusage"
Begin[`implementation`]
All::usage= "someusage"
End[]
EndPackage[]
```
sets the usage message for ``System`All`` and ``System`None``. So, in resume, Mathics is handling scoping in the same way that WMA.